src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -207,51 +331,78 @@ public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document
             uint docCookie;
             if (RunningDocumentTable.TryGetCookieForInitializedDocument(document.Key.Moniker, out docCookie))
             {
-                TryProcessOpenForDocCookie(docCookie);
+                TryProcessOpenForDocCookie(docCookie, cancellationToken);
             }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            RemovePendingDocumentInitializationTask(document);
         }
 
-        private void TryProcessOpenForDocCookie(uint docCookie)
+        private void TryProcessOpenForDocCookie(uint docCookie, CancellationToken cancellationToken)
         {
+            AssertIsForeground();
+
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Or actually: I'd consider that this function shouldn't take a cancellation token. This mutates state, so by the contract it's not clear what cancellation means -- can you end up in some torn state? Did it happen? Did it not happen? But maybe that results in harder to write code, I'm not sure.